### PR TITLE
support external modules

### PR DESCRIPTION
**Pull request by osman**
- both mutators and filters can now be externally defined
 - split up the manager into multiple files
 - clean up code

### CODE REVIEW

File: app/api/mutators/mutators.controller.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+              .then(()=> res.status(HttpStatus.NO_CONTENT).json())
+              .catch((e)=> utils.processError(e, res)),
+
+  /**
+   * Returns list of all mutators in a namespace by id
+   */
+  getAllIdsByNamespace: (req, res) =>
+                            new ObjectId({namespace: res.params.namespace, id: 'dummy'}).validate()
+                            .then((objectId)=> services.mutators.getIds(objectId.namespace))
+                            .then((ids)=> res.status(HttpStatus.OK).json({ids: ids}))
```
Comment (effervescentia) [inline app/api/mutators/mutators.controller.js]: you can just do `{ ids }`

File: app/models/schemas/filters.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+module.exports = {
+  type:       'object',
```
Comment (effervescentia) [inline app/models/schemas/filters.js]: convert to JSON file
`require('filters.json')`

File: app/models/schemas/mutators.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module.exports = {
```
Comment (effervescentia) [inline app/models/schemas/mutators.js]: convert to JSON
`require('mutators.json');`

File: app/services/manager.js [hunk 4]
```diff
@@ -46,345 +34,38 @@ const Manager = function (redisClient) {
...
-      return result;
-    }, {});
-  });
-};
+const getWorkersStatus = () =>
+    purgeOldWorkerData()
+    .then(()=> redis.hgetall(WORKER_STATUS_KEY))
+    .then((workersStatus)=>
+        _.reduce(workersStatus, (result, status, workerName) => {
+          result[workerName] = JSON.parse(status);
```
Comment (effervescentia) [inline app/services/manager.js]: I think you can do this instead
```js
return _.assign(result, { [workerName]: JSON.parse(status) });
```

File: app/services/mutators.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      Promise.mapSeries(mutators.actions, (action)=> {
+        const id     = new ObjectId({namespace: _.isString(action.namespace) ? action.namespace : taskName, id: action.id});
+        id.arguments = action.arguments || mutators.arguments;
+        return id;
+      })
+      .then((objectIds)=> Promise.mapSeries(objectIds, (objectId)=>
+          objectId.validate()
+          .then(()=> redis.hget(getNamespacedKey(objectId.namespace), objectId.id))
+          .then((src)=> {
+            const module     = compiler.compile(src);
```
Comment (effervescentia) [inline app/services/mutators.js]: another case for _.assign

File: app/services/subtasks.js [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+  /**
+   * Pop a job off the queue and return it
+   *
+   * @returns {Promise.<TResult>}
+   */
+  self.fetch = (taskId)=> {
+    return Task.validateId(taskId)
+    .then(()=> redis.lpop(Task.backlogQueueKey(taskId)))
+    .then((subtaskID)=> {
+      log.info("ID: %s", subtaskID);
```
Comment (effervescentia) [inline app/services/subtasks.js]: double quotes

File: app/services/subtasks.js [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+   * @param task
+   * @param allIndices
+   * @param loadedFilters
+   * @returns {*}
+   */
+  self.filterDocumentSubtasks = (task, allIndices, loadedFilters) => {
+    const predicate                = (allFilters)=> (input)=> _.reduce(allFilters, (result, filter)=> result || filter.predicate(input), false);
+    const getTypesFromMappings     = (mappings)=> _.reduce(mappings, (result, type, name)=> {
+      type.name = name;
+      result.push(type);
```
Comment (effervescentia) [inline app/services/subtasks.js]: `return result.concat(type);`

File: app/services/tasks.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+   * @param taskId
+   * @param subtask
+   * @param message
+   * @returns {*|{arity, flags, keyStart, keyStop, step}}
+   */
+  self.logError = (taskId, subtask, message) => {
+    const time_ms = moment().valueOf();
+    const error   = JSON.stringify({
+      subtask: Subtask.coerce(subtask),
+      message: message
```
Comment (effervescentia) [inline app/services/tasks.js]: just `message` (no value)

File: app/services/tasks.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+      Task.validateId(taskId)
+      .then(()=> redis.hgetall(Task.progressKey(taskId)))
+      .then((overallProgress)=>
+          _.reduce(overallProgress, (result, rawProgress, rawSubtask) => {
+            const progress = {
+              subtask:  new Subtask(JSON.parse(rawSubtask)),
+              progress: new Progress(JSON.parse(rawProgress))
+            };
+
+            result.push(progress);
```
Comment (effervescentia) [inline app/services/tasks.js]: `concat`

File: app/services/tasks.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+      .then(() => redis.zrangebyscore(Task.errorKey(taskId), '-inf', '+inf', 'WITHSCORES'))
+      .then((rawErrors)=> {
+        const skip   = 2;
+        const errors = [];
+        for (let i = 0; i < rawErrors.length; i += skip) {
+          const errorInfo     = JSON.parse(rawErrors[i]);
+          errorInfo.subtask   = new Subtask(errorInfo.subtask);
+          errorInfo.timestamp = moment(rawErrors[i + 1], 'x').toISOString();
+
+          errors.push(errorInfo);
```
Comment (effervescentia) [inline app/services/tasks.js]: `concat` + `assign`

File: app/services/transfer.js [hunk 11]
```diff
@@ -381,29 +244,57 @@ const Transfer = function (sourceEs, destEs) {
...
+  self.clearMutators = ()=> {
+    mutators = {};
+  };
+
+  self.setMutators = (newMutators)=> {
+    mutators = newMutators;
+  };
+
+  self.getMutators = ()=> {
+    return _.cloneDeep(mutators);
```
Comment (effervescentia) [inline app/services/transfer.js]: don't need a return (one-liner)

File: app/services/transfer.js [hunk 12]
```diff
@@ -381,29 +244,57 @@ const Transfer = function (sourceEs, destEs) {
     });
   };
 
+  self.clearMutators = ()=> {
+    mutators = {};
+  };
+
+  self.setMutators = (newMutators)=> {
+    mutators = newMutators;
```
Comment (effervescentia) [inline app/services/transfer.js]: oneliner

File: app/services/transfer.js [hunk 13]
```diff
@@ -381,29 +244,57 @@ const Transfer = function (sourceEs, destEs) {
     });
   };
 
+  self.clearMutators = ()=> {
+    mutators = {};
```
Comment (effervescentia) [inline app/services/transfer.js]: oneliner

File: app/services/transfer.js [hunk 14]
```diff
@@ -381,29 +244,57 @@ const Transfer = function (sourceEs, destEs) {
...
-      }
-    });
-    queue.push({
-      doc:           document._source,
-      doc_as_upsert: true
-    });
+  self.mutate = (originals, type) => {
+    if (_.isArray(mutators[type]) && mutators[type].length > 0) {
+      return _.reduce(originals, (result, original)=> {
+        for (let i = 0; i < mutators[type].length; i++) {
```
Comment (effervescentia) [inline app/services/transfer.js]: `mutators[type].forEach()`

File: app/services/transfer.js [hunk 15]
```diff
@@ -381,29 +244,57 @@ const Transfer = function (sourceEs, destEs) {
...
+            const mutated = mutators[type][i].mutate(original, mutators[type][i].arguments);
 
-    return queue;
+            if (!_.isUndefined(mutated) && !_.isNull(mutated) && !_.isEmpty(mutated)) {
+              original = mutated;
+            }
+          }
+        }
+
+        result.push(original);
```
Comment (effervescentia) [inline app/services/transfer.js]: `concat`

File: app/services/transfer.js [hunk 16]
```diff
@@ -465,45 +352,57 @@ const Transfer = function (sourceEs, destEs) {
...
+ * @param targetIndices
+ * @returns {Promise.<TResult>}
+ */
+Transfer.getIndices = (client, targetIndices) =>
+    !_.isString(targetIndices) || targetIndices.length === 0 ?
+        Promise.reject(new Error('targetIndices must be string with length')) :
+        client.indices.get({index: targetIndices, allowNoIndices: true})
+        .then((response)=> _.reduce(response, (result, index, name)=> {
+          log.info('got index: ', name);
+          index.name = name;
```
Comment (effervescentia) [inline app/services/transfer.js]: `assign` and `concat`

File: app/services/transfer.js [hunk 17]
```diff
@@ -465,45 +352,57 @@ const Transfer = function (sourceEs, destEs) {
...
+ * @param client
+ * @param targetTemplates
+ * @returns {Promise.<T>}
+ */
+Transfer.getTemplates = (client, targetTemplates) =>
+    !_.isString(targetTemplates) || targetTemplates.length === 0 ?
+        Promise.reject(new Error('targetTemplates must be string with length')) :
+        client.indices.getTemplate({name: targetTemplates})
+        .then((templates)=> _.reduce(templates, (result, template, name)=> {
+          template.name = name;
```
Comment (effervescentia) [inline app/services/transfer.js]: `assign` and `concat`

File: app/services/worker.js [hunk 18]
```diff
@@ -1,127 +1,130 @@
...
-        if (tasks.length === 0) {
+  const taskIds     = [];
+  const getTaskName = ()=> taskIds.length !== 0 ?
+      taskIds.pop() :
+      tasks.getAll()
+      .then((allTasks)=> {
+        if (allTasks.length === 0) {
           return null;
+        } else {
+          _.forEach(allTasks, (task)=> taskIds.push(task));
```
Comment (effervescentia) [inline app/services/worker.js]: don't need the lodash forEach

File: app/services/worker.js [hunk 19]
```diff
@@ -1,127 +1,130 @@
...
+            if (!subtask) {
+              log.info('No subtask to execute, waiting...');
+              manager.workerHeartbeat(name, {status: `waiting for subtask...`});  // Not waiting for promise
+              return timeoutPromise(RUN_CHECK_INTERVAL_MS);
+            }
+
+            manager.workerHeartbeat(name, {
+              status:  'starting..',
+              task:    taskId,
+              subtask: subtask
```
Comment (effervescentia) [inline app/services/worker.js]: can be subtask

File: app/services/worker.js [hunk 20]
```diff
@@ -131,64 +134,64 @@ const Worker = function (redisClient) {
...
       total:       subtask.count,
       worker:      name
     });
 
-    self.manager.workerHeartbeat(name, {
+    manager.workerHeartbeat(name, {
       status:   'running',
-      task:     taskName,
+      task:     taskId,
       subtask:  subtask,
```
Comment (effervescentia) [inline app/services/worker.js]: can just use the key

File: config/elasticsearch.js [hunk 21]
```diff
@@ -28,12 +31,29 @@ const LogToBunyan = function () {
...
+    uri = `http://${uri}`;
+  }
+
+  const results    = sr('GET', uri);
+  const version    = JSON.parse(results.getBody('utf8')).version.number;
+  const apiVersion = `${semver.major(version)}.${semver.minor(version)}`;
   return new elasticsearch.Client({
-    host:       host,
+    host:       {
+      host: host,
```
Comment (effervescentia) [inline config/elasticsearch.js]: just use the keys

File: config/redis.js [hunk 22]
```diff
@@ -1,12 +1,11 @@
...
-const createClient = (host, port)=>{
-  return new Redis({
-    port: port,
-    host: host,
-    db:   10
-  });
-};
+const createClient = (host, port)=>
+    new Redis({
+      host: host,
```
Comment (effervescentia) [inline config/redis.js]: keys only

File: tests/config.js [hunk 23]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
```
Comment (effervescentia) [inline tests/config.js]: can be a JSON file

File: tests/services/filters.spec.js [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+    utils   = new Utils();
+    done();
+  });
+
+  after((done)=> {
+    return redis.flushdb().finally(done);
+  });
+
+  beforeEach((done)=> {
+    objectId = new ObjectId({namespace: ns, id: id});
```
Comment (effervescentia) [inline tests/services/filters.spec.js]: can use id key

File: tests/services/filters.spec.js [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+      .then(()=> filters.add(objectId, utils.loadFile(`${__dirname}/validFilters/index.js`)))
+      .catch((e) => expect(e.message).to.be.equals(`Filter 'namespace/id' exists, delete first.`))
+      .then(() => done())
+  );
+
+  it('load a specific filter with top-level args', (done)=> {
+    filters.add(objectId, utils.loadFile(`${__dirname}/validFilters/index.js`))
+    .then(()=> filters.load(ns, {
+      actions:   [
+        {id: id}
```
Comment (effervescentia) [inline tests/services/filters.spec.js]: use key directly

File: tests/services/filters.spec.js [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+    })
+    .catch(done);
+  });
+
+  it('load a specific filter with arg overrides', (done)=> {
+    filters.add(objectId, utils.loadFile(`${__dirname}/validFilters/index.js`))
+    .then(()=> filters.load(ns, {
+      actions:   [
+        {
+          id:        id,
```
Comment (effervescentia) [inline tests/services/filters.spec.js]: id key directly

File: tests/services/filters.spec.js [hunk 27]
```diff
@@ -0,0 +1,236 @@
...
+      validateFilter(data);
+      done();
+    })
+    .catch(done);
+  });
+
+  it('load a multiple filters in different namespaces', (done)=> {
+    const objectId1 = new ObjectId({namespace: 'task', id: id});
+    const objectId2 = new ObjectId({namespace: 'global', id: id});
+    const objectId3 = new ObjectId({namespace: 'othernamespace', id: id});
```
Comment (effervescentia) [inline tests/services/filters.spec.js]: use id keys directly

**Comment by osman** [comment]
@ehacke please review

**Pull request merged**